### ISSUE TITLE
[NativeAOT] Invalid IL or CLR metadata + exe crash

### ISSUE THREAD

**Issue by dellamonica** [dotnet/runtimelab#274]
Warning on compilation using Microsoft.DotNet.ILCompiler -v 6.0.0-*

```
> EXEC : warning : Method `[System.Interactive.Async]System.Linq.AsyncEnumerableEx+<>c__DisplayClass6_0`2+<<Catch>g__Core|0>d<System.__Canon,System.__Canon>.MoveNext()` will always throw because: Invalid IL or CLR metadata [(...)\Tests.csproj]
> EXEC : warning : Method `[System.Interactive.Async]System.Linq.AsyncEnumerableEx+<>c__DisplayClass7_0`2+<<Catch>g__Core|0>d<System.__Canon,System.__Canon>.MoveNext()` will always throw because: Invalid IL or CLR metadata [(...)\Tests.csproj]
> EXEC : warning : Method `[System.Interactive.Async]System.Linq.AsyncEnumerableEx+<>c__DisplayClass8_0`2+<<Catch>g__Core|0>d<System.__Canon,System.__Canon>.MoveNext()` will always throw because: Invalid IL or CLR metadata [(...)\Tests.csproj]
```

The package version is
`<PackageReference Include="System.Interactive.Async" Version="4.1.1" />`

The only method that I use from this package is 
` IAsyncEnumerable<TSource> AsyncEnumerableEx.Merge<TSource>(params IAsyncEnumerable<TSource>[] sources);`

After compiling, the application crashes (I suspect that it happens as soon as the await foreach on the IAsyncEnumerable starts because the Console.WriteLine before the enumeration is printed as expected).

**Comment by jkotas** [comment]
This is caused by `catch (TException ...)` at https://github.com/dotnet/reactive/blob/main/Ix.NET/Source/System.Interactive.Async/System/Linq/Operators/Catch.cs#L62.  

The underlying problem is duplicate of dotnet/runtime#67806.

The NativeAOT compiler is running in higher-compatibility mode by default that only trims framework assemblies, and tries to compile everything in non-framework assemblies including 3rd party NuGet packages in case they are accessed by reflection. It means that you are hitting this problem with `Catch` operator even through it is not used by your program at all.

You can fix this problem by specifying `<RootAllApplicationAssemblies>false</RootAllApplicationAssemblies>` in your .csproj file. It will allow trimming of unused code in the applications assemblies. `Catch` operator will not be compiled so you won't see the error above, and also the final binary will be much smaller. 

You can read about more options to optimize size and troubleshoot size at https://github.com/dotnet/corert/tree/master/Documentation/using-corert

**Comment by jkotas** [comment]
Duplicate of dotnet/runtime#67806

**Comment by dellamonica** [comment]
Thanks @jkotas, the .csproj changes worked. 

The final exe was just 13Mb, great job! However, even though I optimized for speed, it seems that the native app was much *slower* than the jitted dotnet app. It is difficult to do a direct comparison because the native app generates different results (any slightly different implementation of say, a collection method, could cause this, I guess), but I believe that the difference is significant.

I did have to change one thing in the code to make it work: an Array.Sort was causing a crash with InvalidOperation_IComparerFailed. Replacing it with MemoryExtensions.Sort for the Span<T> version worked. I didn't find any GitHub Issue mentioning this problem, but if you think it's a good idea I can open a new one, here's the relevant part of the stack trace:

 ```
---> EETypeRva:0x009120F8: InvalidOperation_IComparerFailed
 ---> EETypeRva:0x00910FC8: Argument_ImplementIComparable
StackTrace_AtWordSystem.Collections.Comparer.Compare(Object, Object) + 0x161
StackTrace_AtWordSystem.Collections.Generic.ObjectComparer`1.Compare(T, T) + 0x6b
StackTrace_AtWordSystem.Collections.Generic.ArraySortHelper`1.SwapIfGreater(Span`1, Comparison`1, Int32, Int32) + 0xc5
StackTrace_AtWordSystem.Collections.Generic.ArraySortHelper`1.IntroSort(Span`1, Int32, Comparison`1) + 0x97
StackTrace_AtWordSystem.Collections.Generic.ArraySortHelper`1.IntrospectiveSort(Span`1, Comparison`1) + 0x80
StackTrace_AtWordSystem.Collections.Generic.ArraySortHelper`1.Sort(Span`1, IComparer`1) + 0x78
   Exception_EndOfInnerExceptionStack
StackTrace_AtWordSystem.ThrowHelper.ThrowInvalidOperationException(ExceptionResource, Exception) + 0x51
StackTrace_AtWordSystem.Collections.Generic.ArraySortHelper`1.Sort(Span`1, IComparer`1) + 0xc1
```

**Comment by jkotas** [comment]
If you can isolate this into small repro or collect data on the perf difference, we will be happy to take a look.

If there is signifiant performance difference, it should be easy to tell where it is coming from by looking at the profile (we usually 
 use https://github.com/Microsoft/perfview to profiling). The usual problems are System.Linq.Expressions or System.Text.RegularExpressions that are interpreter for native AOT, but have more efficient code generated for runtime with a JIT.

**Comment by dellamonica** [comment]
There is no usage of Linq in hot paths for this code, and no regex at all.

I've run both JIT and Native versions on the same input with multiple seeds (to average out computation time) and the results were pretty consistent: about **14s** on JIT and about **57s** on Native.

PerfView is a nice tool but I really did not spot anything obvious. Here are some of the things I noticed: 

- There's a Span ClearWithReferences that is taking more CPU time than I would expect, but it accounts for less than 3% of exclusive time.
- MemoryMarshal.Cast is showing up with significant percentage, which is also a surprise.
- Comparing the profiling of the JIT vs Native version there is a lot of discrepancy in the weight (exclusive %) of certain methods.

The source code uses Span<T>, Vector<T> and intrinsics quite a lot.

If you'd like me to try something specific or take a closer look at the profiling data, let me know.

**Comment by jkotas** [comment]
> MemoryMarshal.Cast

MemoryMarshal.Cast is marked with aggresiveinlining. It should not be showing up in the profile at all unless the JIT hits limitations that make it disable optimizations. What are the generic arguments of MemoryMarshal.Cast that shows the discrepancy? Are the methods that it is called from particularly large?

> intrinsics

NativeAOT currently assumes the least common denominator for hardware intrinsics by default. We do not have code multi-versioning that would be required to support optional intrinsics.  If your code depends on Avx2 or other instruction sets that are not part of the least common denominator, you can try including `<IlcArg Include="--instructionset:avx2" />` in your project file. It will make the compiler generate the Avx2 instructions; however the binary is only going to run on machines that have Avx2.

**Comment by dellamonica** [comment]
> MemoryMarshal.Cast is marked with aggresiveinlining. It should not be showing up in the profile at all unless the JIT hits limitations that make it disable optimizations. What are the generic arguments of MemoryMarshal.Cast that shows the discrepancy? Are the methods that it is called from particularly large?

- int32 => uint64: called from a ~70 line method which has loops, intrinsics
- int32 => struct with a few bool/int/enum fields: called from a fairly simple method (~50 lines)

> NativeAOT currently assumes the least common denominator for hardware intrinsics by default. We do not have code multi-versioning that would be required to support optional intrinsics. If your code depends on Avx2 or other instruction sets that are not part of the least common denominator, you can try including `<IlcArg Include="--instructionset:avx2" />` in your project file. It will make the compiler generate the Avx2 instructions; however the binary is only going to run on machines that have Avx2.

I'm using Sse2 and Sse41 (guarded by .IsSupported). With the instructionset as AVX2 (and also with SSE2) I did not see any noticeable improvement in the running time.

**Comment by jkotas** [comment]
> int32 => uint64: called from a ~70 line method which has loops, intrinsics
> int32 => struct with a few bool/int/enum fields: called from a fairly simple method (~50 lines)

Would you mind to share the C# code and the assembly code (you can get it in debugger) for these two methods? If it helps, we can do it over email if you are not able to share this on github. My email address is in my github profile.

**Comment by dellamonica** [comment]
Sent to your e-mail.

**Comment by dellamonica** [comment]
Regarding perf issue and inlining: it was my blunder!

I basically used my unit tests project to give NativeAOT a quick try. In the Test .csproj I had <Optimize>false</Optimize>. Once I set the Optimize to true for NativeAOT compilation, the execution times are now comparable with the Jitted version and Cast was apparently inlined as expected.

Perhaps the only contribution that this can have to the project is either the recommendation to include <Optimize>true</Optimize> explicitly in the documentation or perhaps the NativeAOT chain should just ignore that option? The only reason I disable optimizations (for the Test project) is for an easier experience when debugging the tests. I fail to see why would someone want to produce a native application without compiler optimizations.

**Comment by jkotas** [comment]
Happy to hear that we have figured it out! Thanks.

> why would someone want to produce a native application without compiler optimizations.

For the same reason you have `<Optimize>false</Optimize>` in your project: easier for debugging. I have done it a lot to debug non-codegen related issues.

**Comment by dellamonica** [comment]
> Happy to hear that we have figured it out! Thanks.

Thank you, and sorry for the mix up.

> > why would someone want to produce a native application without compiler optimizations.
> 
> For the same reason you have `<Optimize>false</Optimize>` in your project: easier for debugging. I have done it a lot to debug non-codegen related issues.

Fair enough, although it wouldn't be too bad showing "Generating native code /// Compiler optimizations are turned off" + whatever other info you think might be relevant.
